webknossos/webknossos/dataset/_utils/buffered_slice_writer.py

```diff
@@ -0,0 +1,164 @@
+
+    def _write_buffer(self) -> None:
+        if len(self.buffer) == 0:
+            return
+
+        assert (
+            len(self.buffer) <= self.buffer_size
+        ), "The WKW buffer is larger than the defined batch_size. The buffer should have been flushed earlier. This is probably a bug in the BufferedSliceWriter."
+
+        uniq_dtypes = set(map(lambda _slice: _slice.dtype, self.buffer))
+        assert (
+            len(uniq_dtypes) == 1
+        ), "The buffer of BufferedSliceWriter contains slices with differing dtype."
+        assert uniq_dtypes.pop() == self.dtype, (
+            "The buffer of BufferedSliceWriter contains slices with a dtype "
+            "which differs from the dtype with which the BufferedSliceWriter was instantiated."
+        )
+
+        logging.debug(
+            "({}) Writing {} slices at position {}.".format(
+                getpid(), len(self.buffer), self.buffer_start_slice
+            )
+        )
+        log_memory_consumption()
+
+        try:
+            assert (
+                self.buffer_start_slice is not None
+            ), "Failed to write buffer: The buffer_start_slice is not set."
+            buffer_start = [0, 0, 0]
+            buffer_start[self.dimension] = self.buffer_start_slice
+            offset = cast(
+                Tuple[int, int, int],
+                tuple(
+                    [off + buff_off for off, buff_off in zip(self.offset, buffer_start)]
+                ),
+            )
+            max_width = max(slice.shape[-2] for slice in self.buffer)
+            max_height = max(slice.shape[-1] for slice in self.buffer)
+
+            self.buffer = [
+                np.pad(
+                    slice,
+                    mode="constant",
+                    pad_width=[
+                        (0, 0),
+                        (0, max_width - slice.shape[-2]),
+                        (0, max_height - slice.shape[-1]),
+                    ],
+                )
+                for slice in self.buffer
+            ]
+
+            data = np.concatenate(
+                [np.expand_dims(slice, self.dimension + 1) for slice in self.buffer],
+                axis=self.dimension + 1,
+            )
+            self.view.write(data, offset)
+
+        except Exception as exc:
+            logging.error(
+                "({}) An exception occurred in BufferedSliceWriter._write_buffer with {} "
+                "slices at position {}. Original error is:\n{}:{}\n\nTraceback:".format(
+                    getpid(),
+                    len(self.buffer),
+                    self.buffer_start_slice,
+                    type(exc).__name__,
+                    exc,
+                )
+            )
+            traceback.print_tb(exc.__traceback__)
+            logging.error("\n")
+
+            raise exc
+        finally:
+            self.buffer = []
+
+    def get_slice_generator(self) -> Generator[None, np.ndarray, None]:
```

Comment:
Please also mark this as private:
```suggestion
    def _get_slice_generator(self) -> Generator[None, np.ndarray, None]:
```